### ISSUE TITLE
[java] UseDiamondOperator false negatives

### ISSUE THREAD

**Issue by maikelsteneker** [pmd/pmd#2545]
**Affects PMD Version:** 6.23.0 and 6.24.0 (but not 6.22.0)

Make sure, to test with the latest PMD version.

**Description:**

I've identified a couple of (what I believe are) false negatives for rule UseDiamondOperator. These seem to have been introduced in PMD 6.23.0, probably as a side effect of fixing #1723.

These false negatives are as follows:
* When using nested type parameters (e.g., `List<Map<String,Object>>`), the rule doesn't seem to trigger even in cases where it should
* When using a ternary operator (`boolean ? value : otherValue`), the rule doesn't seem to trigger either
* It also doesn't seem to work in a static initializer

**Code Sample demonstrating the issue:**

```
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
public class File {
  List<Map<String,Object>> l = new ArrayList<Map<String,Object>>(); // FN
  public void test() {
    final List<String> l2;
    l2 = true ? new ArrayList<String>() : new ArrayList<String>(); // FN
  }
  static {
    l = new ArrayList<Map<String,Object>>(); // FN
  }
}
```
PMD 6.22.0 reports the following:

> /tmp/File.java:6:	Explicit type arguments can be replaced by Diamond Operator
/tmp/File.java:9:	Explicit type arguments can be replaced by Diamond Operator
/tmp/File.java:9:	Explicit type arguments can be replaced by Diamond Operator
/tmp/File.java:12:	Explicit type arguments can be replaced by Diamond Operator

PMD 6.23.0 and 6.24.0 report no violations at all.

**Steps to reproduce:**

Please provide detailed steps for how we can reproduce the bug.

1. Save the code example above to /tmp/File.java
2.  Run `./run.sh pmd -d /tmp/File.java -R category/java/codestyle.xml/UseDiamondOperator` 

**Running PMD through:** *CLI*

**Comment by adangel** [comment]
That's a bit trickier: The nested type parameters are clearly a false-negative. The cases with the ternary operator are ok for Java7 and false-negatives for Java8+. In Java8, the type inference has been improved.
The static case is just another usage of nested type parameters.

I'm preparing a PR, which introduces a property for the rule, to retain Java7 compatibility (which is disabled by default) and reports these cases as well.

**Comment by maikelsteneker** [comment]
@adangel Thank you for the investigation and improvements. I didn't realize the subtle difference between Java 7 and Java 8, but I'm happy to see both cases have been accounted for!
